src/core/ucc_context.c

```diff
@@ -204,3 +190,100 @@ void ucc_context_config_print(const ucc_context_config_h config, FILE *stream,
             config->lib->full_prefix, (ucc_config_print_flags_t)flags);
     }
 }
+
+static inline void ucc_copy_context_params(ucc_context_params_t *dst,
+                                           const ucc_context_params_t *src)
+{
+    dst->mask = src->mask;
+    UCC_COPY_PARAM_BY_FIELD(dst, src, UCC_CONTEXT_PARAM_FIELD_TYPE, ctx_type);
+    UCC_COPY_PARAM_BY_FIELD(dst, src, UCC_CONTEXT_PARAM_FIELD_COLL_OOB, oob);
+    UCC_COPY_PARAM_BY_FIELD(dst, src, UCC_CONTEXT_PARAM_FIELD_ID, ctx_id);
+    UCC_COPY_PARAM_BY_FIELD(dst, src, UCC_CONTEXT_PARAM_FIELD_COLL_SYNC_TYPE,
+                            sync_type);
+}
+
+ucc_status_t ucc_context_create(ucc_lib_h lib,
+                                const ucc_context_params_t *params,
+                                const ucc_context_config_h  config,
+                                ucc_context_h *context)
+{
+    ucc_base_context_params_t b_params;
+    ucc_base_context_t       *b_ctx;
+    ucc_cl_lib_t             *cl_lib;
+    ucc_context_t            *ctx;
+    ucc_status_t              status;
+    uint64_t                  i;
+    int                       num_cls;
+
+    num_cls = config->n_cl_cfg;
+    ctx     = ucc_malloc(sizeof(ucc_context_t), "ucc_context");
+    if (!ctx) {
+        ucc_error("failed to allocate %zd bytes for ucc_context",
+                  sizeof(ucc_context_t));
+        status = UCC_ERR_NO_MEMORY;
+        goto error;
+    }
+    ctx->lib = lib;
+    ucc_copy_context_params(&ctx->params, params);
+    ucc_copy_context_params(&b_params.params, params);
```

Comment:
Do you need to copy? Do you expect the structures to change in future?

Reply:
There are 2 copies here: 1st - we cache user input on ucc_context_t. Main reason: there is OOB in params, that we will need. 2nd copy: to pass params down into CL/TL interfaces we need to pack it into "base_params" according to iface. THose copies are super-tiny and hence super fast and in the control path. should be ok i think.